how-does-fix-with-copilot-work.md

```diff
@@ -0,0 +1,168 @@
+          maxLength={LINE_CONTEXT_MAX_SIZE} 
+          documentContext={documentContext} />
+        <DiagnosticRelatedInfo 
+          diagnostic={d} 
+          cookbook={cookbook} 
+          document={documentContext.document} />
+        <DiagnosticSuggestedFix cookbook={cookbook} />
+    </>;
+})
+```
+
+What components make up the prompt? The main ones:
+
+- Include code surrounding the error.
+- Format the error nicely.
+- Provide instructions to Copilot.
+
+### Include Surrounding Code
+
+By default *Fix With Copilot* includes the entire file around the error. Of course, most of this time the entire file won't fit in the prompt. In that case, it includes the code in current function and summarizes code outside. 
+
+The algorithm to do this is complicated. It uses the parse tree to assign depths and distances to each block in the file. For example, the top-level `interface Node<T>` in the diagram is `distance: 2, depth: 1` from the error-containing function `errorInHere`. The body of `findRoot<T>` is `distance: 1, depth: 2`, since it's only one function above the error, but it's one down from the top level.
+
+![Code on the left surrounded by nested blocks. Each block is with with numbers. Block with `export interface Node<T>` annotated `{ distance: 2, depth: 1 }`. Block with `payload: T; parent?: Node<T>;` annotated `2,2`. Block with `export function findRoot<T>(start: Node<T>)` annotated `1,1`. Block with `let node: Node<T> | undefined = start; let prev: Node<T> | undefined = node; while (node)` annotated `1,2`. Block with `prev = node; node = node!.parent;` annotated `1,3`. Block with `function errorInHere() { // error is in this function }` annotated `0,1`. The same code is on the right, but here, any block with numbers that sum to 4 or more has had its code replaced with `// ...`.](images/ai-summarised-document.png)
+
+Once every block has a distance and a depth, each one gets a score and everything above a certain score gets abbreviated. To keep the example simple, let's just add `distance` and `depth` to get the score. Then you can see that anything with a scor above 3 is abbreviated. The body of `while (node)` was `distance:1 + depth:3` and the body of `interface Node<T>` was `distance:2 + depth:2`, so both get abbreviated.
+
+### Format Error Message
+
+Formatting the error starts with the error text itself, then adds a couple of sub-components: `DiagnosticRelatedInfo` and `DiagnosticSuggestedFix`.
+
+`DiagnosticRelatedInfo` itself adds two kinds of information: the first is any related code that the error itself contains. For example, the TypeScript error  *"Duplicate declaration"* includes the duplicated declaration as related info. The second kind of information is the source of imported functions in calls with bad argument types. *Fix With Copilot* finds the source of a called function in a language-independent way, using VS Code's language service API, although right now the code only applies to TypeScript.
```

Comment:
This paragraph should expand on how source is retrieved.  If the code is small enough, include a code sample of tree sitter and language service usage.